### PR TITLE
Scavengers: Support HTML in hunts

### PR DESCRIPTION
**Pull request by Lucas-Meijer**
As discussed with Computerwizard8800, aQrator, and Scavengers roomstaff.

Implementation of HTML for Scavengers hunts, similar to HTML in viewrepeats.
Also adds maximum size for Scavengers hunts, to avoid images being too large.

Worked together with Alexander489

### CODE REVIEW

**Comment by aQrator** [comment]
Confirming approval

File: server/chat-plugins/mafia.ts [hunk 1]
```diff
@@ -2,13 +2,14 @@ import {Utils, FS} from '../../lib';
 
 interface MafiaData {
 	// keys for all of these are IDs
-	alignments: {[k: string]: MafiaDataAlignment};
-	roles: {[k: string]: MafiaDataRole};
-	themes: {[k: string]: MafiaDataTheme};
-	IDEAs: {[k: string]: MafiaDataIDEA};
-	terms: {[k: string]: MafiaDataTerm};
-	aliases: {[k: string]: ID};
+	alignments: { [k: string]: MafiaDataAlignment };
```
Comment (KrisXV) [inline server/chat-plugins/mafia.ts:5]: can you undo all of these

File: server/chat-plugins/scavengers.ts [hunk 2]
```diff
@@ -1410,7 +1432,7 @@ const ScavengerCommands: Chat.ChatCommands = {
 		const res = ScavengerHunt.parseQuestions(params);
 		if (res.err) return this.errorReply(res.err);
 
-		room.game = new ScavengerHunt(room, user, hosts, gameType, res.result, mod);
+		room.game = new ScavengerHunt(room, user, hosts, gameType, res.result, isHTML, mod);
```
Comment (PartMan7) [inline server/chat-plugins/scavengers.ts:1435]: The class is starting to have a bit too many arguments; could you make this a destructured parameter instead?

File: server/chat-plugins/scavengers.ts [hunk 3]
```diff
@@ -362,6 +365,8 @@ export class ScavengerHunt extends Rooms.RoomGame<ScavengerHuntPlayer> {
 
 		this.hosts = hosts;
 
+		this.isHTML = isHTML ? isHTML : false;
```
Comment (PartMan7) [inline server/chat-plugins/scavengers.ts:368]: Any reason why this isn't just `this.isHTML = isHTML` or `this.isHTML = !!isHTML`?

File: server/chat-plugins/scavengers.ts [hunk 4]
```diff
@@ -1736,6 +1765,7 @@ const ScavengerCommands: Chat.ChatCommands = {
 			room.settings.scavQueue.push({
 				hosts: next.hosts,
 				questions: correctlyFormattedQuestions,
+				isHTML: this.cmd.includes('html'),
```
Comment (PartMan7) [inline server/chat-plugins/scavengers.ts:1768]: If the isHTML from line 1744 is still in scope, probably reuse that here

**Comment by PartMan7** [comment]
Could you also check the cases in `scavenger-games.ts` where we send HTML? `onShowEndBoard` for Minesweeper is an example that would need to be handled.

File: server/chat-plugins/scavengers.ts [hunk 5]
```diff
@@ -443,8 +449,8 @@ export class ScavengerHunt extends Rooms.RoomGame<ScavengerHuntPlayer> {
 		const huntType = `${article} ${newHunt ? 'new ' : ''}${this.gameType}`;
 
 		return `|raw|<div class="broadcast-blue"><strong>${huntType} scavenger hunt by <em>${hosts}</em> has been started${staffHost}.</strong>` +
-			`<div style="border:1px solid #CCC;padding:4px 6px;margin:4px 1px">` +
-			`<strong><em>Hint #1:</em> ${Chat.formatText(this.questions[0].hint)}</strong>` +
+			`<div style="border:1px solid #CCC;padding:4px 6px;margin:4px 1px; overflow:scroll; max-height: 50vh">` +
```
Comment (PartMan7) [inline server/chat-plugins/scavengers.ts:452]: Could you add a similar %-based value for max-width? Currently, images overflow horizontally (you've already handled the vertical cases).

File: server/chat-plugins/scavenger-games.ts [hunk 6]
```diff
@@ -781,11 +781,11 @@ const MODES: {[k: string]: GameMode | string} = {
 							if (staffHost) staffHost.sendTo(this.room, `${targetUser.name} has received their first hint early.`);
 							targetUser.sendTo(
 								this.room,
-								`|raw|<strong>The first hint to the next hunt is:</strong> ${Chat.formatText(this.questions[0].hint)}`
+								`|raw|<strong>The first hint to the next hunt is:</strong> ${this.isHTML ? this.questions[0].hint : Chat.formatText(this.questions[0].hint)}`
```
Comment (PartMan7) [inline server/chat-plugins/scavenger-games.ts:784]: Would wrap handling be needed for these?

File: server/chat-plugins/scavengers.ts [hunk 7]
```diff
@@ -238,7 +239,7 @@ function formatQueue(queue: QueuedHunt[] | undefined, viewer: User, room: Room,
 	} else {
 		buffer = `<tr><td colspan=3>The scavenger queue is currently empty.</td></tr>`;
 	}
-	let template = `<div class="ladder"><table style="width: 100%"><tr><th>By</th><th>Questions</th></tr>${showStaff ? buffer : buffer.replace(/<button.*?>.+?<\/button>/gi, '')}</table></div>`;
+	let template = `<div class="ladder" style="overflow:scroll; max-height: 50vh"><table style="width: 100%"><tr><th>By</th><th>Questions</th></tr>${showStaff ? buffer : buffer.replace(/<button.*?>.+?<\/button>/gi, '')}</table></div>`;
```
Comment (PartMan7) [inline server/chat-plugins/scavengers.ts:242]: Re: both /scav queue and /viewhunt, could you make the wrap apply on each row as opposed to the full table? If Q1 has an overflowing image, the other questions are shoved out of view.

**Review by PartMan7 — CHANGES_REQUESTED** [review]
Whoops, sorry this took me a while to get to.

**Review by PartMan7 — APPROVED** [review]
Looks great; thanks!

File: server/chat-plugins/scavengers.ts [hunk 8]
```diff
@@ -443,8 +449,8 @@ export class ScavengerHunt extends Rooms.RoomGame<ScavengerHuntPlayer> {
 		const huntType = `${article} ${newHunt ? 'new ' : ''}${this.gameType}`;
 
 		return `|raw|<div class="broadcast-blue"><strong>${huntType} scavenger hunt by <em>${hosts}</em> has been started${staffHost}.</strong>` +
-			`<div style="border:1px solid #CCC;padding:4px 6px;margin:4px 1px">` +
-			`<strong><em>Hint #1:</em> ${Chat.formatText(this.questions[0].hint)}</strong>` +
+			`<div style="border:1px solid #CCC;padding:4px 6px;margin:4px 1px; overflow:scroll; max-height: 50vh">` +
+			`<strong><em>Hint #1:</em> ${this.isHTML ? this.questions[0].hint : Chat.formatText(this.questions[0].hint)}</strong>` +
```
Comment (HoeenCoder) [inline server/chat-plugins/scavengers.ts:453]: I see a large number of very similar statements in the form of `this.isHTML ? text : Chat.formatText(text)`. Have you considered modularizing this? More specifically make a method such as:
```typescript
formatOutput(text: string): string {
	return this.isHTML ? text : Chat.formatText(text);
}
```

I will mention I do see some of these statements using Utils.escapeHTML instead, evaluate which method you want to use within the proposed method.

**Review by HoeenCoder — COMMENTED** [review]
Excellent work overall, one thing I noticed to consider.

**Review by mia-pi-git — APPROVED** [review]
Make the change Hoeen suggested and tag me when you're done, I'll merge this. LGTM otherwise.

**Comment by Lucas-Meijer** [comment]
> Make the change Hoeen suggested and tag me when you're done, I'll merge this. LGTM otherwise.

@mia-pi-git it's ready!

File: server/chat-plugins/scavengers.ts [hunk 9]
```diff
@@ -226,19 +227,19 @@ function formatQueue(queue: QueuedHunt[] | undefined, viewer: User, room: Room,
 							return Utils.html`<span style="color: green"><em>[${q.join(' / ')}]</em></span><br />`;
 						} else {
 							q = q as string;
-							return Utils.escapeHTML(q);
+							return item.isHTML ? q : Utils.escapeHTML(q);
```
Comment (mia-pi-git) [inline server/chat-plugins/scavengers.ts:230]:
```suggestion
							return item.formatOutput(q);
```

Comment (mia-pi-git) [reply server/chat-plugins/scavengers.ts:230]: Oh, actually, you can ignore this. Went and poked at it a bit to be sure.

**Pull request merged**